sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala

```diff
@@ -361,3 +361,78 @@ class JacksonParser(
     }
   }
 }
+
+object JacksonParser {
+  private[spark] def splitDocuments(input: InputStream) = new Iterator[String] {
+
+    private implicit class JsonCharacter(char: Char) {
+      def isJsonObjectFinished(endToken: Option[Char]): Boolean = {
+        endToken match {
+          case None => char == '}' || char == ']'
+          case Some(x) => char == x
+        }
+      }
+    }
+    private var currentChar: Char = input.read().toChar
+    private var previousToken: Option[Char] = None
+    private var nextRecord = readNext
+
+    override def hasNext: Boolean = nextRecord.isDefined
+
+    override def next(): String = {
+      if (!hasNext) {
+        throw new NoSuchElementException("End of stream")
+      }
+      val curRecord = nextRecord.get
+      nextRecord = readNext
+      curRecord
+    }
+
+    private def moveToNextChar() = {
+      if (!currentChar.isWhitespace) {
+        previousToken = Some(currentChar)
+      }
+      currentChar = input.read().toChar
+    }
+
+    private def readJsonObject: Option[String] = {
+      val endToken = currentChar match {
+        case '{' => Some('}')
+        case '[' => Some(']')
+        case _ => None
+      }
+
+      val sb = new StringBuilder()
+      sb.append(currentChar)
+      while (!currentChar.isJsonObjectFinished(endToken) && input.available() > 0) {
+        moveToNextChar()
+        currentChar match {
+          case '{' | '[' =>
```

Comment:
yes, you are right. Unfortunately it is very hard to support properly also permissive mode. I am closing this PR. I'll open a new one if I'll manage to find a better way.

Reply:
It is quite sad as you probably had a lot of work in it. I hope you can reuse some parts in the next try.

Reply:
the main issue is that I should subclass the Jackson parsers to make it working with the current approach, since when a JSON is not valid Jackson doesn't allow to get the string which caused the problem. And we are using many. Thus it's gonna be a very big effort. If you have better ideas, please feel free to submit a PR. Thanks.

Reply:
The problem I presented I think can ne be solved with extending your solution, but as I see you could have more problems in your mind. It would be nice to collect all the cases where the splitDocuments should work and how. Do you have time for this? Then maybe we can find the solution. I would be glad to help by reviewing and thinking together or solving parts of the problem.

Reply:
yes, but then then escapes should be taken in account, etc. etc. Then we would nearly have to rewrite the jackson library logic, which is something not desirable. This is the reason why I said that. Yes sure, I would be happy if we can find a solution together, but I think it is hard with this approach.